## 바이오컴퓨팅 과제
Bio-Computing Assignment

Python을 이용하여 과제 수행. 

----------------
## 과제1
[Assignment 1](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment1.py) 

String Manipulation - Reverse Complement

+ **Purpose**

Getting familiar with python scripts

Practice of string data manipulation

----------------
## 과제2
[Assignment 2](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment2.py) 

String Manipulation - GC Content

+ **Purpose**

Getting familiar with python scripts

Practice of string data manipulation

----------------
## 과제3 
[Assignment 3](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment3.py)

String Manipulation - Repeats 

+ **Purpose**
  
Getting familiar with python scripts

Practice of string data manipulation

---------------
## 과제4
[Assignment 4](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment4.py)

Edit Distance Measurement
 
+ **Purpose**
  
Getting familiar with python scripts

Practice of measuring sequence similarities

----------------
## 과제5
[Assignment 5](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment5.py) 

Local Pairwise Sequence Alignment

+ **Purpose**

Understanding of dynamic programming for sequence alignment

Practice of implementing pairwise sequence alignment

----------------
## 과제6
[Assignment 6](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment6.py)
Multiple Sequence Alignment

+ **Purpose**
  
Understanding of multiple sequence alignment

Practice of implementing heuristics to solve an intractable problem


+ Star Alignment Algorithm

1. For each sequence x,

2. Sum the scores of all pairwise global alignments against x.
 
3. Choose the center sequence that maximizes the score sum.

4. For each of the other sequences x, where x is not the center sequence,
   
5. Determine the optimal global alignment between x and the center sequence.

6. Merge the pairwise alignment into the multiple sequence alignment.

Your code should read a single file of multiple protein sequences in a FASTA format, print the multiple alignment result to an output file, and print the elapsed time of your code in microseconds to the screen. In the last line, mark "*" on the columns containing identical amino acid residues across all sequences. To score the alignment, use a BLOSUM62 substitution matrix for matches and mismatches, and use -5 as the gap penalty. Your code will take the input file and output file names as command-line arguments.

----------------
## 과제7 
[Assignment 7](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment7.py) 

Pattern Finding

+ **Purpose**

Practice of pattern finding in python

Practice of implementing an exhaustive algorithm

----------------
## 과제8
[Assignment 8](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment8.py) 

Approximate Pattern Finding

+ **Purpose**

Practice of implementing the actual pattern finding problem in python

Practice of implementing an exhaustive algorithm

----------------
## 과제 9
[Assignment 9](https://github.com/ysungJ/Assignment/blob/main/Bio-Computing/Assignment9.py) 

Character-Based Evolutionary Tree Reconstruction

+ **Purpose**

Practice of understanding the small parsimony problem

Practice of implementing the character-based evolutionary tree reconstruction algorithm

----------------

